src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -411,41 +398,126 @@ void GcsPlacementGroupScheduler::OnAllBundleSchedulingRequestReturned(
-    node_to_bundles->emplace(iter.first, iter.second.size());
+  for (const auto &node_it : alive_nodes) {
+    const auto &node_id = node_it.first;
+    const auto &bundle_locations_on_node = bundle_location_index_.Get(node_id);
+    RAY_CHECK(bundle_locations_on_node)
+        << "Bundle locations haven't been registered for node id " << node_id;
+    const int bundles_size = bundle_locations_on_node.value()->size();
+    node_to_bundles->emplace(node_id, bundles_size);
   }
 
-  std::shared_ptr<BundleLocations> bundle_locations = nullptr;
-  auto iter = placement_group_to_bundle_locations_.find(placement_group_id);
-  if (iter != placement_group_to_bundle_locations_.end()) {
-    bundle_locations = iter->second;
-  }
+  auto &bundle_locations = bundle_location_index_.Get(placement_group_id);
   return std::unique_ptr<ScheduleContext>(new ScheduleContext(
       std::move(node_to_bundles), bundle_locations, gcs_node_manager_));
 }
 
 absl::flat_hash_map<PlacementGroupID, std::vector<int64_t>>
 GcsPlacementGroupScheduler::GetBundlesOnNode(const ClientID &node_id) {
   absl::flat_hash_map<PlacementGroupID, std::vector<int64_t>> bundles_on_node;
-  const auto node_iter = node_to_leased_bundles_.find(node_id);
-  if (node_iter != node_to_leased_bundles_.end()) {
-    const auto &bundles = node_iter->second;
-    for (auto &bundle : bundles) {
-      bundles_on_node[bundle.first.first].push_back(bundle.second->BundleId().second);
+  const auto &maybe_bundle_locations = bundle_location_index_.Get(node_id);
+  if (maybe_bundle_locations.has_value()) {
+    const auto &bundle_locations = maybe_bundle_locations.value();
+    for (auto &bundle : *bundle_locations) {
+      const auto &bundle_placement_group_id = bundle.first.first;
+      const auto &bundle_index = bundle.first.second;
+      bundles_on_node[bundle_placement_group_id].push_back(bundle_index);
     }
-    node_to_leased_bundles_.erase(node_iter);
+    bundle_location_index_.Erase(node_id);
   }
   return bundles_on_node;
 }
 
+void BundleLocationIndex::AddBundleLocations(
+    const PlacementGroupID &placement_group_id,
+    std::shared_ptr<BundleLocations> bundle_locations) {
+  placement_group_to_bundle_locations_.emplace(placement_group_id, bundle_locations);
+  for (const auto &iter : *bundle_locations) {
+    const auto &location = iter.second;
+    const auto &node_id = location.first;
+    const auto &bundle_sepc = location.second;
+    node_to_leased_bundles_[location.first]->emplace(
+        bundle_sepc->BundleId(), std::make_pair(node_id, bundle_sepc));
+  }
+}
+
+bool BundleLocationIndex::Erase(const ClientID &node_id) {
+  const auto leased_bundles_it = node_to_leased_bundles_.find(node_id);
+  if (leased_bundles_it == node_to_leased_bundles_.end()) {
+    return false;
+  }
+
+  const auto &bundle_locations = leased_bundles_it->second;
+  for (const auto &bundle_location : *bundle_locations) {
+    // Remove corresponding placement group id.
```

Comment:
Sorry, I didn't find the code.

Reply:
No worries haha. I was surprised I couldn't find this problem from the PR before!